sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java

```diff
@@ -442,12 +441,29 @@ public static Mono<BinaryData> fromFlux(Flux<ByteBuffer> data, Long length, bool
                     MAX_ARRAY_SIZE)));
         }
         if (bufferContent) {
-            if (length != null) {
-                return FluxUtil.collectBytesInByteBufferStream(data, length.intValue())
-                    .flatMap(bytes -> Mono.just(BinaryData.fromBytes(bytes)));
-            }
-            return FluxUtil.collectBytesInByteBufferStream(data)
-                .flatMap(bytes -> Mono.just(BinaryData.fromBytes(bytes)));
+            // Buffer the Flux<ByteBuffer> content using ByteBuffers of equal size to the original ByteBuffer.
+            // Previously this was using FluxUtil.collectBytesInByteBufferStream which runs into two issues:
+            //
+            // 1. The content is limited in size as it collects into a byte array which is limited to ~2GB in size.
+            // 2. This could lead to a very large chunk of data existing which can cause pauses when allocating large
+            //    arrays.
+            long[] trueLength = new long[]{0};
+            return data.map(buffer -> {
+                int bufferSize = buffer.remaining();
+                ByteBuffer copy = ByteBuffer.allocate(bufferSize);
+                trueLength[0] += bufferSize;
+                copy.put(buffer);
+                copy.flip();
+
+                return copy;
+            })
+            .collect(LinkedList::new, (BiConsumer<LinkedList<ByteBuffer>, ByteBuffer>) LinkedList::add)
+            .map(buffers -> {
+                // TODO (alzimmer): What should be done when length != null but it differs from the true length
```

Comment:
log warning or error or, if you're confident it's invalid, then throw?

Reply:
For now I'm going to leave the comment and current behavior and do a wider investigation to what calls this to see if changing this would drastically change runtime behavior